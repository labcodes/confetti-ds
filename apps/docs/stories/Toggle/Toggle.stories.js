import React from "react";

import { Toggle as Component } from "../../../../packages/confetti-ds/src";

export default {
  title: "Toggle/Toggle",
  component: Component,
};

export const Toggle = (args) => <Component {...args} />;
Toggle.args = {
  id: "demo-toggle",
  name: "demo-toggle",
};
