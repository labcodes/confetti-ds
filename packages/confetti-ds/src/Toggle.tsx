import React, {SyntheticEvent, useState, useEffect } from "react";
import { isUndefined } from "lodash";

interface TooggleProps {
  /** Text that will serve as unique identifier. It's also an important accessibility tool. */
  id?: string;
  /** Text that will specify the HTML `name` attribute of an `<input>` element. */
  name?: string;
  /** Defines a static value for the Toggle. If set, internal logic is deactivated. */
  value: boolean;
  /** Defines the Toggle color. */
  color: "teal" | "purple";
  /** Disables the Toggle. Will be read by screen readers. When true, will override `disabled`. */
  ariaDisabled: boolean;
  /** Disables the Toggle. Won't be read by screen readers. */
  disabled: boolean;
  /** Sets value to true by default. */
  defaultValue: boolean;
  /** Action executed when the Toggle is clicked. */
  handleToggle: (event?: SyntheticEvent) => any;
}

export default function Toggle({
  defaultValue,
  id,
  color = "teal",
  name,
  disabled = false,
  ariaDisabled = false,
  value,
  handleToggle,
}: TooggleProps) {
  const [localValue, setLocalValue] = useState(
    !isUndefined(defaultValue) ? defaultValue : false
  );
  const isChecked = !isUndefined(value) ? value : localValue;

  useEffect(() => {
    if (!isUndefined(defaultValue) && !isUndefined(value)) {
      // eslint-disable-next-line no-console
      console.warn(
        `You are setting both value and defaultValue for input ${name} at the same time. We always initialize the toggle with defaultValue. Make sure this is the behaviour you want.`
      );
    }
  }, []);

  const handleOnChange = (event) => {
    if (!isUndefined(handleToggle)) {
      handleToggle(event);
    }
    setLocalValue(!localValue);
  };

  return (
    <label
      className={`
          lab-toggle
          ${color ? `lab-toggle--${color}` : ""}
          ${disabled || ariaDisabled ? " lab-toggle--disabled" : ""}
        `}
      htmlFor={name}
    >
      <input
        type="checkbox"
        id={id}
        name={name}
        className="lab-toggle__input"
        checked={isChecked}
        onChange={!ariaDisabled ? handleOnChange : () => {}}
        disabled={(!ariaDisabled && disabled) || undefined}
        aria-disabled={ariaDisabled || undefined}
      />
      <span className="lab-toggle__slider" />
    </label>
  );
}
