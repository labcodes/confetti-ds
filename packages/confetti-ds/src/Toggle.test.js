/* eslint-disable no-console */
import React from "react";
import { mount, shallow } from "enzyme";
import renderer from "react-test-renderer";

import Toggle from "./Toggle";

describe("Toggle", () => {
  it("renders with base props", async () => {
    const component = <Toggle id="test-toggle" name="test-toggle" />;
    expect(mount(component)).toBeTruthy();
    const renderedComponent = renderer.create(component).toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders as expected when passing disabled as true", async () => {
    const renderedComponent = renderer
      .create(<Toggle id="test-toggle" name="test-toggle" disabled />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders as expected when passing a different color", async () => {
    const renderedComponent = renderer
      .create(
        <Toggle id="test-toggle" color="purple" name="test-toggle" disabled />
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("initializes toggled if defaultValue is passed", async () => {
    const mountedComponent = mount(
      <Toggle id="test-toggle" name="test-toggle" defaultValue />
    );
    expect(mountedComponent.find({ checked: true })).toHaveLength(1);
  });

  it("initializes untoggled if defaultValue is passed as false", async () => {
    const mountedComponent = mount(
      <Toggle id="test-toggle" name="test-toggle" defaultValue={false} />
    );
    expect(mountedComponent.find({ checked: false })).toHaveLength(1);
  });

  it("initializes untoggled if defaultValue is not passed", async () => {
    const mountedComponent = mount(
      <Toggle id="test-toggle" name="test-toggle" />
    );
    expect(mountedComponent.find({ checked: false })).toHaveLength(1);
  });

  it("changes state when input changes", async () => {
    const mountedComponent = mount(
      <Toggle id="test-toggle" name="test-toggle" />
    );

    expect(mountedComponent.find({ checked: false })).toHaveLength(1);
    mountedComponent.find("input").at(0).simulate("change");
    expect(mountedComponent.find({ checked: true })).toHaveLength(1);
  });

  it("calls props.handleToggle passing event when input changes", async () => {
    const mockHandleToggle = jest.fn();
    const mountedComponent = mount(
      <Toggle
        id="test-toggle"
        name="test-toggle"
        handleToggle={mockHandleToggle}
      />
    );

    expect(mountedComponent.find({ checked: false })).toHaveLength(1);
    expect(mockHandleToggle).not.toBeCalled();

    mountedComponent.find("input").at(0).simulate("change", { test: "event" });

    expect(mountedComponent.find({ checked: true })).toHaveLength(1);
    expect(mockHandleToggle).toBeCalledWith(
      expect.objectContaining({ test: "event" })
    );
  });

  it("doesn't trigger onChange if ariaDisabled", async () => {
    const mockHandleToggle = jest.fn();
    const shallowComponent = shallow(
      <Toggle
        ariaDisabled
        handleToggle={mockHandleToggle}
        id="test-toggle"
        name="test-toggle"
      />
    );

    expect(shallowComponent.find({ checked: false })).toHaveLength(1);
    expect(mockHandleToggle).not.toBeCalled();

    shallowComponent.find("input").at(0).simulate("change", { test: "event" });

    expect(shallowComponent.find({ checked: false })).toHaveLength(1);
    expect(mockHandleToggle).not.toBeCalled();
  });
});
