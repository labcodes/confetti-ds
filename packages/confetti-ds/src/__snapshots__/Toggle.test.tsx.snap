// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle renders as expected when passing a different color 1`] = `
<label
  className="
          lab-toggle
          lab-toggle--purple
           lab-toggle--disabled
        "
  htmlFor="test-toggle"
>
  <input
    checked={false}
    className="lab-toggle__input"
    disabled={true}
    id="test-toggle"
    name="test-toggle"
    onChange={[Function]}
    type="checkbox"
  />
  <span
    className="lab-toggle__slider"
  />
</label>
`;

exports[`Toggle renders as expected when passing disabled as true 1`] = `
<label
  className="
          lab-toggle
          lab-toggle--teal
           lab-toggle--disabled
        "
  htmlFor="test-toggle"
>
  <input
    checked={false}
    className="lab-toggle__input"
    disabled={true}
    id="test-toggle"
    name="test-toggle"
    onChange={[Function]}
    type="checkbox"
  />
  <span
    className="lab-toggle__slider"
  />
</label>
`;

exports[`Toggle renders with base props 1`] = `
<label
  className="
          lab-toggle
          lab-toggle--teal
          
        "
  htmlFor="test-toggle"
>
  <input
    checked={false}
    className="lab-toggle__input"
    id="test-toggle"
    name="test-toggle"
    onChange={[Function]}
    type="checkbox"
  />
  <span
    className="lab-toggle__slider"
  />
</label>
`;
